gen/qr.hoon

```diff
@@ -0,0 +1,359 @@
+::
+::  TODO type this better, will ease implementation of new features
+::
+++  grid  (list (list ?))
+::
+::::::::::::::::::::::::::::::::::::::::::::::::3: main logic
+::
+++  pretty-print
+  |=  qr=grid
+  ^-  tang
+  =/  rq  (flop qr)  :: to get around `tang` printing weirdness
+  %+  turn  rq
+  |=  row/(list ?)
+  :-  %leaf
+  ^-  tape
+  %+  turn  row
+  |=  solid/?
+  ?:  solid
+    ' '
+  '█'
+::
+++  encode
+  |=  message=tape
+  ^-  grid
+  %-  add-border
+  %-  add-mask-information
+  %-  pick-mask
+  %-  add-data  
+  :-  %-  add-error-correction 
+          message
+  %-  create-base
+    "5L"
+::
+::::::::::::::::::::::::::::::::::::::::::::::::4: qr formatting helpers
+::
+++  create-base
+  |=  version=tape
+  ^-  grid
+  ?.  =(version "5L")
+    ~&  [%fatal-error "sorry, only QR version 5L is currently supported"]
+    !!
+  %-  atm  
+  :-  ;:  weld 
+    dark-module
+    timing-pattern 
+    alignment-pattern
+    finder-pattern
+  ==
+::
+  %-  init-qr-base
+  37  :: hardcoding QR version 5-L
+::
+++  init-qr-base
+  |=  n=@u
+  ^-  grid
+  %+  reap  n     
+  %+  reap  (mul 2 n)  :: one '█' is too narrow, use two together for each "module" 
+                       :: TODO yikes module abstraction needed. let pretty-print do this?
+      %.n
+::
+++  add-border
+  |=  qr=grid
+  ^-  grid
+  =/  n=@u  (lent qr)
+  =/  i=@u  0
+  :-  (reap (mul 2 (add 2 n)) %.n)
+  |-
+  ?:  =(i n)
+    :-  (reap (mul 2 (add 2 n)) %.n)  ~
+  :-  (zing [[%.n %.n ~] (snag i qr) [%.n %.n ~] ~])
+  $(i (add 1 i))
+::
+++  add-mask-information
+  |=  [mask=@u qr=grid]
+  ^-  grid
+  =/  info  (format-strings mask)
+  =/  i=@u  0  :: TODO dont hardcode mask 000
+  |-
+  ?:  =(i (lent info))
+    qr
+  ?:  =((snag i info) 1)
+    =/  coords-one  (snag i info-route-one)
+    =/  coords-two  (snag i info-route-two)
+    %-  tm  :-  [-.coords-one +.coords-one]
+    %-  tm  :-  [-.coords-two +.coords-two]
+    $(i (add 1 i))
+  $(i (add 1 i))
+::
+++  pick-mask
+  |=  qr=grid
+  ^-  [@u grid]
+  =/  i=@u  0
+  :-  0  :: TODO hardcoding mask 000 now. implement other masks + penalty scoring
+  |-
+  ?:  =(i (lent data-route))
+    qr
+  =/  coords  (snag i data-route)
+  =/  x  -.coords
+  =/  y  +.coords
+  ?:  =(0 (mod (add x y) 2))  :: mask 000
+    %-  tm  :-  [x y]
+    $(i (add 1 i))
+  $(i (add 1 i))
+::
+
+++  atm  :: TODO surely there's a cleaner way to do this too? (+turn...)
+  |=  [coords=(list [@u @u]) b=grid]
+  ^-  grid
+  |-
+  ?~  coords  b
+  %+  tm  i.coords
+  $(coords t.coords)
+::
+++  tm  :: TODO surely there's a cleaner way to do this?
+  |=  [coord=[@u @u] b=grid]
+  ^-  grid
+  =/  c=@u  -.coord
+  =/  r=@u  +.coord
+  =/  count=@u  0
+  =/  s=@u  (mul 2 c)  :: TODO yikes module abstraction needed
+  |-
+  ?:  =(count (lent b))
+    ~
+  =/  row=(list ?)  (snag count b)
+  :-  ?:  =(count r)
+         (zing ~[(scag s row) ~[!(snag s row) !(snag (add 1 s) row)] (slag (add 2 s) row)])
```

Comment:
You might be able to use `:~` here. It'll still be a somewhat complicated structure, but at least slightly easier to parse than one long line.

```hoon
:~  (scag s row)
    :~  !(snag s row)
        !(snag (add 1 s) row)
    ==
    (slag (add 2 s) row)
==
```

Reply:
(I guess if it were actually 1 bit per square you could also do the map thing; turning the map back into a list of lists of characters sounds annoying though, and is kind of most of what `++tm` is already doing)